### ISSUE TITLE
Support most (all?) Scratch blocks

### ISSUE THREAD

**Issue by PullJosh** [PullJosh/scratch-js#18]
# Built-in
## Motion
- [x] Move steps: `this.move(10)`
- [x] Turn right: `this.direction += 15`
- [x] Turn left: `this.direction -= 15`
- [ ] Go to [menu]
- [x] Go to x y: `this.goto(0, 0)`
- [ ] Glide secs to [menu]
- [x] Glide secs to x y: `yield* this.glide(secs, x, y)`
- [x] Point in direction: `this.direction = 90`
- [ ] Point towards
- [x] Change x by: `this.x += 10`
- [x] Set x to: `this.x = 0`
- [x] Change y by: `this.y += 10`
- [x] Set y to: `this.y = 0`
- [ ] If on edge, bounce
- [ ] Set rotation style
- [x] x position: `this.x`
- [x] y position: `this.y`
- [x] direction: `this.direction`

## Looks
- [ ] Say for secs
- [ ] Say
- [ ] Think for secs
- [ ] Think
- [x] Switch costume to: `this.costume = "name"` or `this.costume = 3`
- [x] Next costume: `this.costume++`
- [ ] Switch backdrop to
- [ ] Next backdrop
- [x] Change size by: `this.size += 10`
- [x] Set size to: `this.size = 150`
- [ ] Change effect by
- [ ] Set effect to
  - [ ] Color
  - [ ] Fisheye
  - [ ] Whirl
  - [ ] Pixelate
  - [ ] Mosaic
  - [ ] Brightness
  - [ ] Ghost
- [ ] Clear graphic effects
- [x] show: `this.visible = true`
- [x] hide: `this.visible = false`
- [ ] go to front/back layer
- [ ] go forward/backward layers
- [x] costume number/name: `this.costumeNumber` or `this.costume`
- [ ] backdrop number/name
- [x] size: `this.size`

## Sound
- [x] Play sound until done: `yield* this.playSound("https://example.com/test.mp3")`
- [x] Play sound: `this.playSound("https://example.com/test.mp3")`
- [x] Stop all sounds: `this.stopAllSounds()`
- [ ] Change effect by
- [ ] Set effect to
  - [ ] Pitch
  - [ ] Pan
- [ ] Clear sound effects
- [ ] Change volume by
- [ ] Set volume to
- [ ] volume

## Events
- [x] When green flag clicked: `Trigger.GREEN_FLAG`
- [x] When key pressed: `Trigger.KEY_PRESSED, { key: "space" }`
- [ ] When this sprite clicked
- [ ] When backdrop switches to
- [ ] When greater than
- [x] When I receive: `Trigger.BROADCAST, { name: 'message1' }`
- [x] Broadcast: `this.broadcast("message1")`
- [x] Broadcast and wait: `yield* this.broadcastAndWait("message1")` (does this need to be a separate method from `broadcast`?)

## Control
- [x] Wait secs: `yield* this.wait(1)`
- [x] Repeat: `for (let i = 0; i < 10; i++) { /* do stuff */ yield; }`
- [x] Forever: `while (true) { /* do stuff */ yield; }`
- [x] If: `if () {}`
- [x] If else: `if () {} else {}`
- [x] Wait until: `while(!condition) { yield }`
- [x] Repeat until: `while(!condition) { /* do stuff */ yield; }`
- [ ] Stop
- [ ] When I start as a clone
- [ ] Create clone of
- [ ] Delete this clone

## Sensing
- [ ] Touching
  - [ ] Mouse
  - [ ] Edge
  - [x] Sprite: `this.touching("sprite1")` (plus, as a bonus, fast mode is available for bounding-box checks: `this.touching("sprite1", true)`)
- [ ] Touching color
- [ ] Color is touching color
- [ ] Distance to
- [ ] Ask and wait
- [ ] Answer
- [x] Key pressed: `this.keyPressed("space")`
- [ ] Mouse down
- [x] Mouse x: `this.mouse.x`
- [x] Mouse y: `this.mouse.y`
- [ ] Set drag mode
- [ ] Loudness
- [x] Timer: `this.timer`
- [x] Reset timer: `this.restartTimer()`
- [ ] [property] of [object]
- [x] Current
  - [x] Year: `(new Date().getFullYear())`
  - [x] Month: `(new Date().getMonth() + 1)`
  - [x] Date: `(new Date().getDate())`
  - [x] Day of week: `(new Date().getDay() + 1)`
  - [x] Hour: `(new Date().getHours())`
  - [x] Minute: `(new Date().getMinutes())`
  - [x] Second: `(new Date().getSeconds())`
- [x] Days since 2000: `(((new Date().getTime() - new Date(2000, 0, 1)) / 1000 / 60 + new Date().getTimezoneOffset()) / 60 / 24)` -- This one's pretty beefy. Should it be built-in to the library?
- [x] ~Username~ (N/A)

## Operators
- [x] Add: `2 + 1`
- [x] Subtract: `2 - 1`
- [x] Multiply: `2 * 1`
- [x] Divide: `2 / 1`
- [ ] Pick random: `this.random(1, 10)` -- Still need to replicate Scratch rounding!
- [x] Greater than: `10 > 5`
- [x] Less than: `10 < 5`
- [x] Equal to: `10 == 5` is pretty close to Scratch behavior
- [x] And: `a && b`
- [x] Or: `a || b`
- [x] Not: `!a`
- [x] Join `"hello" + "world"`
- [x] Letter of: `"world"[3]`
- [x] Length of: `"world".length`
- [x] Contains: `"Hello, world!".includes("world")`
- [x] Mod: `10 % 3`
- [x] Round: `Math.round(3.7)`
- [x] Math "of"
  - [x] abs: `Math.abs(-5)`
  - [x] floor: `Math.floor(3.7)`
  - [x] ceiling: `Math.ceil(3.7)`
  - [x] sqrt: `Math.sqrt(16)`
  - [x] sin: `Math.sin(this.scratchToRad(45))`
  - [x] cos: `Math.cos(this.scratchToRad(45))`
  - [x] tan: `Math.tan(this.scratchToRad(45))`
  - [x] asin: `this.radToScratch(Math.asin(0.5))`
  - [x] acos: `this.radToScratch(Math.acos(0.5))`
  - [x] atan: `this.radToScratch(Math.atan(0.5))`
  - [x] ln: `Math.log(5)`
  - [x] log: `Math.log10(5)`
  - [x] e ^: `Math.E ** 5`
  - [x] 10 ^: `10 ** 5`

## Variables
Variables and lists are currently implemented, but in a pretty terrible way. API design needs to be reconsidered.
### Variables
- [ ] Set variable to
- [ ] Change variable by
- [ ] Show variable
- [ ] Hide variable

### Lists
- [ ] Add to list
- [ ] Delete of list
- [ ] Delete all of list
- [ ] Insert at position of list
- [ ] Replace item of list
- [ ] Item of list
- [ ] Item # of string in list
- [ ] Length of list
- [ ] List contains item
- [ ] Show list
- [ ] Hide list

## My Blocks
Custom blocks are a feature. There aren't any predefined blocks to support, because... well. You know.

# Extensions
## Music
- [ ] Play drum for beats
- [ ] Rest for beats
- [ ] Play note for beats
- [ ] Set instrument
- [ ] Set tempo
- [ ] Change tempo
- [ ] Tempo

## Pen
- [x] Erase all: `this.clearPen()`
- [x] Stamp: `this.stamp()`
- [x] Pen down: `this.penDown = true`
- [x] Pen up: `this.penDown = false`
- [x] Set pen color: `this.penColor = "#ff0000"` or `this.penColor = "rgba(255, 0, 0)"` or [any other valid color](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/strokeStyle)
- [ ] Change pen [property]
- [ ] Set pen [property]
- [x] Change pen size: `this.penSize += 10`
- [x] Set pen size: `this.penSize = 10`

## Other Extensions...
Supporting hardware/api scratch extensions natively is out of the scope of scratch-js. However, it may be worth considering the possibility of supporting loading in any generic Scratch 3.0 extension directly. Not sure how difficult it would be.

**Comment by micahlt** [comment]
Looks like you’ve done a lot recently!  Loads of compatibility.  :D  Looks like I can keep writing documentation!

**Comment by PullJosh** [comment]
@micahlt That would be fantastic! :)

**Comment by PullJosh** [comment]
Many blocks are straightforward to implement, but a few pose questions:
* Say and ask: Do we implement proper speech bubbles, or rely on `alert()` and `propmpt()`?
* How do we handle blocks with sprite-selection dropdowns? We can...
  1. Pass a string with the sprite name
  2. Import the sprite class and pass that

  Option 1 is easier (and makes code generation with [`sb-edit`](https://github.com/PullJosh/sb-edit) simpler too), but option 2 less prone to bugs and might even allow smart IDEs to give errors when a sprite's name changes, is deleted, etc...

**Comment by nanaian** [comment]
> Do we implement proper speech bubbles
Yes.

> How do we handle blocks with sprite-selection dropdowns?
Import the sprite class! Magic strings are
* bad
* not good
* confusing, because we've just learnt that a 'string' is just text - why does it suddenly refer to Cat? Cat != "Cat"!
* very bad

You could also provide a `String.prototype.toSprite` or something, like

```js
import Cat from './cat.mjs'

createCloneOf("Cat".toSprite())
createCloneOf(Cat)
```

**Comment by PullJosh** [comment]
@nanaian Agree on speech bubbles.

You're right that magic strings are a bad plan. Thanks for making it clear.

One thing that's kind of weird about passing the class directly is that it *feels* like we should be passing the instance instead. If I run `this.goto(Dog)`, I expect to go to the coordinates of a specific instance of the Dog sprite. Although Scratch has a clear hierarchy with one "parent" sprite and many clones, it seems unnecessarily restrictive to carry this idea over to scratch-js.

**Comment by nanaian** [comment]
> it seems unnecessarily restrictive to carry this idea over to scratch-js

So something more like
```js
const meowth = new Cat() // ie. not automagically instanciated by Project
meowth.myVariable // 'original'
meowth.myVariable = 'somethingElse'

const persian = clone(meowth) // inherits variables of meowth
meowth.myVariable // 'somethingElse'

const skitty = new Cat()
skitty.myVariable // 'original'
```
is on the table?

If we did the scope thing currently being discussed in #19, we would likely encourage doing `global.meowth = meowth` etc to share instances globally. More dissimilar to Scratch, but... better?

**Comment by PullJosh** [comment]
Yeah, that's totally an option. It definitely feels like there could be a nice two-in-one solution to both the scope problem and the accessing instances problem. I just don't know what it is. :P

**Comment by PullJosh** [comment]
I've been testing the [compiler](https://github.com/PullJosh/sb-edit) with a bunch of projects pulled from the Scratch "explore" page (and some from the home page). It seems like the most commonly used features which are still missing are, in roughly decending order, the following:
1. The touching blocks (a sprite, the mouse, or a color)
2. When this sprite clicked (we get this almost for free after adding a "touching mouse" option)
3. Sprite layering
4. Clones
5. Maybe graphic effects next?

**Comment by callowaysutton** [comment]
Using the [Tune.js](https://github.com/Tonejs/Tone.js) library would make adding drums, sound effects, and general audio a lot easier but I think it would mess with the whole concept of not having to build/install anything :/

**Comment by towerofnix** [comment]
Sound blocks can all be marked complete now!

**Comment by PullJosh** [comment]
> Sound blocks can all be marked complete now!

Good call! I just [updated](https://github.com/PullJosh/scratch-js-website/commit/7775c0dc8ed090380b6d365c365e394e2aa24336) the [docs](https://scratch-js.now.sh/docs/sound) as well.

**Comment by towerofnix** [comment]
>I just updated the docs as well.

Oh nice! I was just procrastinating on figuring out how the site repo works and making a PR for that, lol. (Yess!!)

**Comment by apple502j** [comment]
-  If on edge, bounce: #59 
- Touching edge: #59 
- Loudness: #60 
- When backdrop switches to: #58 
-  When greater than: #60

**Comment by BoomerScratch** [comment]
(wow this issue is inactive for a pretty long time I guess)
Scratch's soundbank is available [here](/LLK/scratch-vm/tree/develop/src/extensions/scratch3_music/assets). I think we can use the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) for adding pitch and speed effects to the instruments/drums.
The for ... beats is basically something like the wait ... seconds block (unless you are using instruments, and play it for 0 beats).
If it's still hard to do, maybe take a look at [the source code of the music extension](/LLK/scratch-vm/blob/develop/src/extensions/scratch3_music/index.js)?

**Comment by HanClinto** [comment]
* If on edge, bounce: #114 
* Touching edge: #114

**Comment by HanClinto** [comment]
Cloud variables could be supported with this (extremely simple) Javascript library that seems to be very Scratch-inspired:
https://kihtrak.com/cloud_variable/

Looking through the author's other projects, it appears that he's a lifelong Scratcher. :)

The code for the client is [quite minimalistic](https://github.com/KihtrakRaknas/cloud_variable/blob/master/cloudify.js), and the [server code](https://github.com/KihtrakRaknas/cloud_variable-server/blob/master/server.js) isn't much larger.  We could either link to this project or copy / adapt it more wholesale -- there is no license attached to the code, so we may want to inquire about it.
